README.md
# React Bootstrap Slider

This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).  It was then ejected due to jQuery issue (see below).

## Overview

A ReactJS wrapper for [seiyria's Bootstrap Slider component](https://github.com/seiyria/bootstrap-slider) using Bootstrap.  From version 3.0.0, we are supporting Bootstrap 4 only.  Please use version 2.2.3 if you are still using Bootstrap 3.


## How to install

Install from npm with:

    npm install --save react-bootstrap-slider

or

    yarn add react-bootstrap-slider

Import like so for ES6:

    import ReactBootstrapSlider from 'react-bootstrap-slider';


The control is implemented in UMD format, so should also work for AMD/RequireJS, but I've not tested that.


### Peer Dependencies

React, React Dom and Bootstrap are listed as peer dependencies (peerDependencies section in package.json) for react-bootstrap-slider.  They will _not_ be installed automatically when you install this component into your own project.  You will need to install them yourself, as part of that project, if you have not done so already.  This command will install them for you if you don't yet have them:

    npm install react react-dom prop-types bootstrap@^4 --save
or

    yarn add react react-dom prop-types bootstrap@^4

#### jQuery

The bootstrap-slider component - and, therefore, react-bootstrap-slider - will work with jQuery if it detects it in your project setup, but it is _not_ a requirement.  It works fine without jQuery.  However, if you are using Webpack or Browserify to build your project, you may get a "missing dependency" build error if jQuery is not present.  This is a known, upstream issue in bootstrap-slider.  Please see [How do I exclude the optional JQuery dependency from my build?](https://github.com/seiyria/bootstrap-slider#how-do-i-exclude-the-optional-jquery-dependency-from-my-build) on the Bootstrap Slider's Readme for how you can workaround this issue.  You can also see how I did it by inspecting the /examples/config/webpack.config.js file and searching for "jQuery".

You must also ensure that you have included bootstrap-slider's CSS file in your build, otherwise the control will be blank!  You'll also need Bootstrap's own CSS file too, of course.  If you're using Webpack, which you will be if you have a Create React App based project, then you can import the CSS file directly into your build, like so:

```JavaScript
import "bootstrap/dist/css/bootstrap.css";
import "bootstrap-slider/dist/css/bootstrap-slider.css"
```

Those two imports need to be in your top-level JavaScript file, e.g. App.js.

Or you can simply add the files as links in your HTML file as a link e.g.:

```HTML
<link rel="stylesheet" href="//stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css" />
    <script src="//cdnjs.cloudflare.com/ajax/libs/bootstrap-slider/11.0.2/bootstrap-slider.min.js"></script>
```

checking that the versions match those of Bootstrap and Bootsrap Slider themselves.

## How to use react-bootstrap-slider

Here's an example of how you might call it in your ReactJS's render method:

```JavaScript
<ReactBootstrapSlider
    value={this.state.currentValue}
    change={this.changeValue}
    slideStop={this.changeValue}
    step={this.state.step}
    max={this.state.max}
    min={this.state.min}
    orientation="vertical"
    reversed={true}
    disabled="disabled" />
```

The **value**, **step**, **max** and **min** parameters should be self-explanatory.  The `change` event is the callback method that will be called when the slider actually changes.  (NB: this is the `change` event for the Bootstrap slider.)  As of version 1.1.2, there is a new `slideStop` event, which only triggers when you've _finished_ moving the slider (i.e. you release it).  This makes `slideStop` a deal less "chatty" than the `change` event.  Decide which one you need; you'll only ever need one or the other (unlike my example code above!).

If the optional parameter **disabled** is included, and is set to "disabled", then the slider will display in a disabled state.  If the parameter is not included, or is set to anything else except "disabled", then the slider control will be enabled.

The default **orientation** value is 'horizontal', so you only need to include that prop if you want to set it to 'vertical'.

If **value** is a two member array, then the slider will have two handles: minimum and maximum.  You should also pass win a **range** parameter set to true in this case.

As of version 1.0.5, other values passed into the component will also passed into the underlying bootstrap-slider component via the [Object Spread Operator that's been proposed for ES2017](https://github.com/Microsoft/TypeScript/issues/7617) (yes, that's how cutting edge I am!!).  See [seyria's documentation](https://github.com/seiyria/bootstrap-slider) for a full list of these.  Those parameters have not all been tested yet but the `reversed` parameter definitely works, and I have included this in the demo of the vertical slider.

Version 1.0.6 fixed an issue that prevented ticks props from being rendered.  Pass in ticks props like so:

```JavaScript
ticks = {[0, 100, 200, 300, 400]}
ticks_labels = {["$0", "$100", "$200", "$300", "$400"]}
ticks_snap_bounds = { 30 }
```

CHANGELOG.md
3.0.0 / 2020-06-18
========================
* Bootstrap 4 support.  Thanks to [Dave Bendit](https://github.com/DBendit) for getting the ball rolling on this.
* Split README.md into 3 files, adding CHANGELOG.md and CONTRIBUTING.md.

2.2.3 / 2020-01-25
========================

* Control now transpiled to ES5.  Should fix errors on IE 11 etc.

2.2.2 / 2019-03-25
========================
* Fixed absolute require bug regression in 2.2.0/2.2.1.  (It's @babel/react preset, not @babel/react-app !)

2.2.0 and 2.2.1 / 2019-02-23
========================
* Update to bootstrap-slider 10.
* Major overhaul of Demo vs Component source code.  They're now split into separate folders with separate built steps and package.json files.
* 2.2.1 is a Readme correctly only.  No actual code changes.

2.1.5 / 2018-05-02
========================
* Fixed issue where control was permanently disabled if initial `max` value was set to zero.

2.1.4 / 2018-04-13
========================
* Explicitly allow labelledby to be passed into the control 

2.1.2 & 2.1.3 / 2017-10-21
========================
* 2.1.2 Fixed issue where setting the "disabled" prop did not disable the component on load.
* 2.1.3 Removed a debugging console.log that I introduced in 2.1.2 (doh!).

2.1.1 / 2017-10-03
========================
* Fixed missing peerDependencies and devDependencies issue.

2.1.0 / 2017-09-30
========================
* Updated for React 16 (Fiber).
* Updated dev dependencies to latest versions too.
* Removed React and Bootstrap from dependencies list.  They are now in devDependencies and peerDependencies.

2.0.1 / 2017-08-19
========================
* Module is now destroyed correctly via componentWillUnmount method (thanks to KevBelisle for the PR).

2.0.0 / 2017-04-30
========================
* Updated for React 15.5.x and 16 (Fiber) compatibility by using prop-types package.  This version will not work with older versions of React, hence the major version bump.

1.1.7 / 2017-04-02
========================
* Removed jQuery as any kind of dependency.  It's not actually needed, but it is used by bootstrap-slider if present.

1.1.6 / 2017-03-28
========================
* Moved bootstrap, jquery, react and react-dom from dependencies to peerDependencies and devDependencies. (They need to in both because neither yarn or npm install command actually installs peerDependencies.)

1.1.5 / 2017-03-07
========================
* Remove unnecessary console.log() calls (thanks to saevarom for the PR).
* Fixed path to testBuild script in package.json

1.1.4 / 2017-02-11
========================
* Removed es6bindall as a dependency.  This allows using react-bootstrap-slider with Require.JS.  (PR merge from TalAter.)
* Fixed readme.md error: `reverse` parameter should actually be `reversed`.  (Thanks to chunkiat82 for the pick up.)

1.1.3 / 2016-11-19
========================
* Fixed bug where min, max and step values were not updating correctly when component received new props for them.
* Updated dependencies to latest versions.

1.1.2 / 2016-09-24
========================
* Reverted change of slider trigger event from `change` to `slideStop` in v1.1.1.  This was a dumb way of doing it, and was a borderline breaking change to boot.
* Added new `slideStop` event, which maps to the bootstrap-slider event of the same name
* Added new `change` event, which maps to the bootstrap-slider event of the same name.  Note: this is in _addition_ to the previous `handleChange` event, which will continue to work and is _not_ being deprecated here.

1.1.1: 2016-09-23 
========================
Bad version.  Do not use!  See 1.1.2 comments above.
* Changed main slider trigger event from `change` to `slideStop`.  Component should now be less chatty with improved performance.

1.1.0 / 2016-09-21
========================
* Updated to seiyria's 9.2 version.  This includes the rangeHighlights feature.
* Updated readme to specify inclusion of bootstrap-slider's CSS file.
* Removed reports folder from NPM package

1.0.6 / 2016-08-01.
========================
* Fixed issue where ticks props were not being rendered.  This is because the code was assuming that there would always be min and max props, but the ticks props actually override these.

1.0.5 / 2016-06-01
========================
* Explained how use dual slider controls.  Added test for same.
* Cleaned up demo to code to have one Demo class instead of three.

src/js/app.js
/* eslint-env browser */

import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import ReactBootstrapSlider from "./react-bootstrap-slider.js";

const wrapperDivStyles = {
  backgroundColor: "#E0E0E0",
  padding: "20px",
  width: "300px"
};

const DemoSingleValueSpan = ({ id, value }) => (
  <span>
    Value: <span id={`valueSpan${id}`}>{value}</span>
  </span>
);

DemoSingleValueSpan.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.number.isRequired
};

const DemoMultiValueSpan = ({ id, value }) => (
  <div>
    Lower Value: <span id={`valueSpan${id}Low`}>{value[0]}</span>
    <br />
    Upper Value: <span id={`valueSpan${id}High`}>{value[1]}</span>
    <br />
  </div>
);

DemoMultiValueSpan.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired
};

class Demo extends React.Component {
  constructor(props) {
    super(props);
    // es6BindAll(this, ["changeValue", "changeAxes"]);

    this.state = {
      ...this.props,
      currentValue: props.startValue
    };
    //   delete this.state.startValue;
  }

  changeValue = e => {
    // console.log("changeValue triggered");
    this.setState({ currentValue: e.target.value });
  };

  changeAxes = () => {
    this.setState({
      currentValue: 500,
      min: 0,
      max: 2000,
      step: 100
    });
  };

  render() {
    const newValue = this.state.currentValue;
    const { changeAxesEnabled, id } = this.props;
    let sliderControl, valueSpan, changeAxesButton;
    if (Array.isArray(newValue)) {
      sliderControl = (
        <ReactBootstrapSlider
          {...this.state}
          value={this.state.currentValue}
          change={this.changeValue}
        />
      );
      valueSpan = <DemoMultiValueSpan id={id} value={newValue} />;
    } else {
      sliderControl = (
        <ReactBootstrapSlider
          {...this.state}
          value={this.state.currentValue}
          slideStop={this.changeValue}
        />
      );
      valueSpan = <DemoSingleValueSpan id={id} value={newValue} />;
      changeAxesButton = changeAxesEnabled && (
        <button id={`but${id}`} onClick={this.changeAxes}>
          {" "}
          Change axes{" "}
        </button>
      );
    }
    return (
      <div>
        <div style={wrapperDivStyles}>{sliderControl}</div>
        <br /> <br />
        {valueSpan}
        <br />
        <br />
        {changeAxesButton}
      </div>
    );
  }
}

Demo.propTypes = {
  id: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.number.isRequired).isRequired
  ]),
  startValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.number.isRequired).isRequired
  ]),
  changeAxesEnabled: PropTypes.bool
};

ReactDOM.render(
  <div>
    <div className="demoWrapper">
      <h3>Horizontal (default) demo</h3>
      <Demo
        id="horizontalSlider"
        name="horizontalSliderName"
        startValue={3000}
        max={20000}
        min={1000}
        step={1000}
        tooltip="always"
        changeAxesEnabled={true}
      />
    </div>
    <div className="demoWrapper">
      <h3>Vertical Demo</h3>
      <Demo
        startValue={3000}
        id="verticalSlider"
        orientation="vertical"
        max={20000}
        min={1000}
        step={1000}
        reversed={true}
        changeAxesEnabled={true}
      />
    </div>
    <div className="demoWrapper">
      <h3>Dual demo</h3>
      <Demo
        startValue={[3000, 10000]}
        range={true}
        id="dualSlider"
        max={20000}
        min={1000}
        step={1000}
      />
    </div>
    <div className="demoWrapper">
      <h3>Ticks Demo</h3>
      <Demo
        id="ticksSlider"
        startValue={200}
        ticks={[0, 100, 200, 300, 400]}
        ticks_labels={["$0", "$100", "$200", "$300", "$400"]}
        ticks_snap_bounds={30}
      />
    </div>
    <div className="demoWrapper">
      <h3>Range Highlights demo</h3>
      <Demo
        id="rangeHighlightsSlider"
        startValue={3000}
        max={20000}
        min={1000}
        step={1000}
        rangeHighlights={[
          { start: 1000, end: 5000 },
          { start: 12000, end: 17000 }
        ]}
      />
    </div>
    <div className="demoWrapper">
      <h3>Disabled demo</h3>
      <Demo
        id="disabledSlider"
        name="disabledSliderName"
        startValue={3000}
        max={20000}
        min={1000}
        step={1000}
        disabled="disabled"
        tooltip="always"
      />
    </div>
    <h3>Everything starts at zero demo</h3>
    <Demo
      id="startZeroSlider"
      name="startZeroSliderName"
      startValue={0}
      max={0}
      min={0}
      step={0}
      tooltip="always"
      changeAxesEnabled={true}
    />
  </div>,
  document.getElementById("main")
);
